Make route patterns allow multibyte sub-patterns.

markstory opened the pull request:
This allows route parameter matching patterns to contain multibyte expressions which is useful for restricting parameters in non-ascii character sets.

Refs #8332

markstory committed:
Make route patterns allow multibyte sub-patterns.

This allows route parameter matching patterns to contain multibyte
expressions which is useful for restricting parameters in non-ascii
character sets.

Refs #8332

chinpei215 commented:
+1. But wouldn't it be better if this change was moved to 3.3 or was implemented as an option?
I doubt that everyone knows what `u`  modifier changes mean.
```php
var_dump(preg_match('/^\w$/', 'あ')); // int(0)
var_dump(preg_match('/^\s$/', '　')); // int(0)

var_dump(preg_match('/^\w$/u', 'あ')); // int(1)
var_dump(preg_match('/^\s$/u', '　')); // int(1)
```

ADmad commented:
Having the use of `u` modifier configurable would be nice. Most people don't require multi-byte route element matching.

markstory commented:
How would someone opt-in to multi-byte patterns @ADmad @chinpei215 ?

markstory commented:
I'm fine with targeting this change to 3.3 for what it's worth.

ADmad commented:
@markstory By using `'multibytePattern' => true` in route options perhaps?

chinpei215 commented:
And/or, by using something like `Router::defaultRegexpModifier('ui')`?

jippi commented:
:+1: for configuration option :) multibyte matching is slower, not much, but a bit, meaning route-heavy apps will be slower

markstory commented:
@ADmad I can add that.

markstory committed:
Add multibytePattern option to Route.

Having an option to enable multibyte pattern support lets people opt-in
to more powerful but slower pattern matching. This also reduces the
chance that this new feature/option could cause issues as it is an
opt-in feature.

Refs #8332

markstory commented:
Updated with a `multibytePattern` option.

chinpei215 commented:
+1

ADmad commented:
LGTM

lorenzo commented:
Good job :)

dereuromark commented on src/Routing/Route/Route.php:
```diff
@@ -199,9 +199,13 @@ protected function _writeRoute()
             $this->_greedy = true;
         }
+        $mode = '';
+        if (!empty($this->options['multibytePattern'])) {
```
Would be nice to have class constants for those. Magic string usage from route files is typo friendly.
